pkg/ext/delegate.go

```diff
@@ -0,0 +1,337 @@
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+)
+
+type typeChecker struct {
+	runtime.Object
+}
+
+type typeCheckerList struct {
+	runtime.Object
+}
+
+// XXX: Implement DeleteCollection to simplify everything here
+// var _ rest.StandardStorage = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.Storage = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.Scoper = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.KindProvider = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.GroupVersionKindProvider = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.SingularNameProvider = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+
+var _ rest.Getter = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.Lister = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.GracefulDeleter = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.Creater = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.Updater = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.Watcher = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+
+// Ptr[U] acts as a type constraint such that
+//
+//	T Ptr[U]
+//
+// means that T is a pointer to U and a runtime.Object.
+type Ptr[U any] interface {
+	*U
+	runtime.Object
+}
+
+// delegate is the bridge between k8s.io/apiserver's [rest.Storage] interface and
+// our own Store interface we want developers to use
+//
+// It is used for non-namespaced objects only.
+type delegate[
+	T Ptr[DerefT],
+	DerefT any,
+	TList Ptr[DerefTList],
+	DerefTList any,
+] struct {
+	gvk          schema.GroupVersionKind
+	gvr          schema.GroupVersionResource
+	singularName string
+	store        Store[T, TList]
+	authorizer   authorizer.Authorizer
+}
+
+// New implements [rest.Storage]
+//
+// It uses generics to create the resource and set its GVK.
+func (s *delegate[T, DerefT, TList, DerefTList]) New() runtime.Object {
+	var t DerefT
+	ptrT := T(&t)
+	ptrT.GetObjectKind().SetGroupVersionKind(s.gvk)
+	return ptrT
+}
+
+// Destroy cleans up its resources on shutdown.
+// Destroy has to be implemented in thread-safe way and be prepared
+// for being called more than once.
+func (s *delegate[T, DerefT, TList, DerefTList]) Destroy() {
```

Comment:
From my reading, it is ok to leave empty, but it's up to us. This function is ran when we shutdown or stop the apiserver. It is not meant to delete all resources in the stores.

I have updated the comment to better reflect this.

AS of now, we don't _need_ this, or I can't find a need for it, though it doesn't mean there aren't. Maybe the store would want a Destroy() function to know when they are "shutting down" (although this can be done with a context instead since all request will end up with a context).

Reply:
Thanks, I think that the comment is good enough for now. I will note that this currently does put us in a bit of a bad spot - we could end up leaking resources (i.e. a context) since this doesn't do anything, and an individual store can't override this behavior.